dali/pipeline/operator/builtin/external_source.h

```diff
@@ -167,13 +168,15 @@ class ExternalSource : public Operator<Backend> {
     cv_.notify_one();
   }
 
+
   /**
    * @brief Sets the data that should be passed out of the op
    * on the next iteration.
    */
-  inline void SetDataSource(const vector<Tensor<CPUBackend>> &t) {
+  template<typename SrcBackend>
+  inline void SetDataSource(const vector<Tensor<SrcBackend>> &t, cudaStream_t stream = 0) {
     DALI_ENFORCE(OperatorBase::batch_size_ == static_cast<int>(t.size()),
-      "Data list provided to ExternalSource needs to have batch_size length.");
+                 "Data list provided to ExternalSource needs to have batch_size length.");
     // Note: If we create a GPU source, we will need to figure
     // out what stream we want to do this copy in. CPU we can
     // pass anything as it is ignored.
```

Comment:
If we do that copy on some external stream, and than in RunImpl do another copy out of those buffers on yet another stream, are we sure, that there is no data race?

Don't we need some cross-stream sync here? Can you check it?

I'm afraid of data races.
With CPU and stream 0 there should be no such issues AFAIK, as it is quite synchronous, but when you start using other streams there may be problems.

Reply:
I think the user should sync on the stream he uses to copy. I no stream is provided and the default is used we should sync on our own.
The current assumption is that after you exit from `SetDataSource` data should be safe inside the DALI, so you can do whatever you want with the data source provided to it.
I can imagine you can ask DALI for its stream, work on it and put some even synchronization on it to make sure that `SetDataSource` consumed it. But I don't know how error-prone it could be.